### PR TITLE
Fix pickling dataclasses

### PR DESCRIPTION
**Pull request by pcmoritz**
The metadata field in clsdict['__dataclass_fields__'] was a mappingproxy object, which can't be pickled. This PR converts it to a dict.

### CODE REVIEW

**Comment by ogrisel** [comment]
@pcmoritz @JessicaSchrouff I updated the PR to directly add pickling support for mappingproxy instances. I have also extended the tests a bit. Let me know if you have any comments.

**Comment by pcmoritz** [comment]
Looks good to me, thanks for the update!

**Comment by ogrisel** [comment]
The coverage report is weird. When I compute the coverage locally, I observe similar inconsistencies. codecov is not the culprit, the raw coverage data fails to record the fact that the `_save_mappingproxy` registration is called. This is annoying, I don't understand why it's the case.

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -904,6 +904,12 @@ def save_root_logger(self, obj):
 
     dispatch[logging.RootLogger] = save_root_logger
 
+    if hasattr(types, "MappingProxyType"):
```
Comment (pierreglaser) [inline cloudpickle/cloudpickle.py]: `coverage` has a specific pragma for partial branches:
```suggestion
    if hasattr(types, "MappingProxyType"):   # pragma: no branch
```

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Indeed that seems to work on my box. Thanks your insight. I pushed a new commit to do this systematically for system-based conditions.

**Pull request merged**